File: services/misskey-mixi2-link/src/bridge/filters.ts
import { hasMention } from "./text.js";

export interface MisskeyNoteLike {
  id: string;
  userId: string;
  text: string | null;
  cw?: string | null;
  replyId?: string | null;
  renoteId?: string | null;
  visibility: string;
  files?: { type: string; url: string | null }[];
}

export function imageFiles(note: MisskeyNoteLike): { type: string; url: string }[] {
  return (note.files ?? []).filter(
    (f): f is { type: string; url: string } => f.url != null && f.type.startsWith("image/"),
  );
}

export function shouldForwardNote(
  note: MisskeyNoteLike,
  selfUserId: string,
  mention: string,
): boolean {
  if (note.userId !== selfUserId) return false;
  if (note.replyId != null) return false;
  if (note.renoteId != null) return false;
  if (note.visibility !== "public") return false;
  if (note.cw != null) return false;
  if (!note.text || !hasMention(note.text, mention)) return false;
  return true;
}

File: services/arona/misskey-mixi2-link/bridge/filters.go
package bridge

import "strings"

// MisskeyFile is the subset of a Misskey drive file needed for filtering.
type MisskeyFile struct {
	Type string
	URL  *string
}

// MisskeyNote is the subset of a Misskey note needed for filtering/forwarding.
type MisskeyNote struct {
	ID         string
	UserID     string
	Text       *string
	CW         *string
	ReplyID    *string
	RenoteID   *string
	Visibility string
	Files      []MisskeyFile
}

// ImageFile is a note attachment known to be an image with a usable URL.
type ImageFile struct {
	Type string
	URL  string
}

// ImageFiles returns the note's attachments that have a URL and an image/*
// mime type.
func ImageFiles(note MisskeyNote) []ImageFile {
	var images []ImageFile
	for _, f := range note.Files {
		if f.URL != nil && strings.HasPrefix(f.Type, "image/") {
			images = append(images, ImageFile{Type: f.Type, URL: *f.URL})
		}
	}
	return images
}

// ShouldForwardNote reports whether note should be bridged to mixi2: it must
// be the account owner's own public note, not a reply or renote, without a
// content warning, and must mention the bot.
func ShouldForwardNote(note MisskeyNote, selfUserID string, mention string) bool {
	if note.UserID != selfUserID {
		return false
	}
	if note.ReplyID != nil {
		return false
	}
	if note.RenoteID != nil {
		return false
	}
	if note.Visibility != "public" {
		return false
	}
	if note.CW != nil {
		return false
	}
	if note.Text == nil || *note.Text == "" || !HasMention(*note.Text, mention) {
		return false
	}
	return true
}
